### ISSUE TITLE
Support found date for ALC/UDC

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#12196]
### Tell us your idea!

From support mail:
When setting a lab cache to "Found" it seems not to record the date, when I set it to found. This would be helpful in general and specifically as I export them to GSAK for my record keeping.

### What solution would you suggest?

Save the current date as found date when the user sets an adventure lab cache  manually as "Found".
(Should then also apply to UDC as we also have the found-feature for those)

### Is there an alternative?

Don't implement it. Rare use case?

**Comment by Lineflyer** [comment]
@ztNFny Why thumbs down. Would it interfere with other features or create an inconsistency or alike?

**Comment by fm-sys** [comment]
IMHO we should rather try to switch to the partner API (only for Lab Caches). Then we don't need these ugly workarounds anymore... Or was there a specific reason why we shouldn't do this?

**Comment by Lineflyer** [comment]
I am in favor of it (and if it is just to show to GS that we are willing to work on it if conditions are OK).  Would you @fm-sys be willing to help @baiti with that? I think its too much of a job for a single person.

**Comment by ztNFny** [comment]
Lab API doesn't allow fetching completed labs unfortunately. You can only get completed lab stages (grouped by lab, but you don't know how many stages there are unless you fetch that from another call) or search in an area and get the completion status of the results.

**Comment by Lineflyer** [comment]
@baiti 
Although this is off-topic for this issue. Maybe we should discuss whether we want to implement as @ztNFny offered help as well?
Topic for our next dev call?

**Comment by baiti** [comment]
> Lab API doesn't allow fetching completed labs unfortunately. You can only get completed lab stages (grouped by lab, but you don't know how many stages there are unless you fetch that from another call) or search in an area and get the completion status of the results. I guess we should first implement the authenticated search. As I know the labs API quite well I can help if it's wanted.

Quite a lot can be accomplished via the Lab API. However, our official (and my be only temporary) API key extremely limits the methods we can call.

You weren't with us at the time when our key wasn't "restricted" to what it is now. Initially I already had implemented to retrieve the true found state for a user, then all of a sudden the respective method no longer worked.

**Comment by baiti** [comment]
> @baiti Although this is off-topic for this issue. Maybe we should discuss whether we want to implement as @ztNFny offered help as well? Topic for our next dev call?

Good idea. Looking forward to it.

**Comment by ztNFny** [comment]
I'm not talking sbout the current cgeo API key, what I mentioned are limitations of the API itself.
Getting the state of one or more adventures is no problem, getting a list of all completed just isn't implenented as an endpoint (only started which may be completed)

**Comment by baiti** [comment]
> I'm not talking sbout the current cgeo API key, what I mentioned are limitations of the API itself. Getting the state of one or more adventures is no problem, getting a list of all completed just isn't implenented as an endpoint (only started which may be completed)

`/api/GeocacheLogs`, if used with the right key, gives you all you need except ONE thing: The answers you gave to all those questions. Which is a BIG deficiency in their implementation. The user has no documentation of what he answered. He has to take notes somehow. Frequently those answers are needed in "Bonus caches" and if you haven't taken care of logging them to a piece of paper you're lost as there is no way to retrieve them. Neither on their App nor on the Web.

**Comment by ztNFny** [comment]
> api/GeocacheLogs, if used with the right key, gives you all you need except ONE thing: 

No, that one only returns info about Geocaches (aka stages), not Adventures. So this is adventures where you have completed at least 1 stage. To figure out how many stages there are you must query the Adventure (or keep that info yourself as i do)

**Comment by baiti** [comment]
> > api/GeocacheLogs, if used with the right key, gives you all you need except ONE thing:
> 
> No, that one only returns completion info about Geocaches (aka stages), not Adventures (although grouped by adventure). So this is adventures where you have completed at least 1 stage. To figure out how many stages there are in total / whether you completed the adventure you must query the Adventure (or keep that info (number of stages) yourself as i do in a separate db). From just /GeocacheLogs you won't be able to tell whether it's a completed Adventure with only 3 stages or a started one with 3 of 5 stages completed.

You may need to combine more than one of these api methods to get a full picture but trust me, you can get a full picture. I can't go into any more details on this channel and besides, it doesn't matter either as we do not, and  never will have an authorization from GC to get this data and use it in our own application.

**Comment by ztNFny** [comment]
> You may need to combine more than one of these api methods to get a full picture but trust me, you can get a full picture. I

Of course you can - I've even written above what you need to do. But the fundamental problem for getting found labs specifically is that you need to do a lot of calls to get the full picture.

> it doesn't matter either as we do not, and never will have an authorization from GC to get this data and use it in our own application.

If that's the assumption we can stop the whole discussion about "do we want to implement Labs API" right here?!

**Comment by baiti** [comment]
> > You may need to combine more than one of these api methods to get a full picture but trust me, you can get a full picture. I
> 
> Of course you can - I've even written above what you need to do. But the fundamental problem for getting found labs specifically is that you need to do a lot of calls to get the full picture.

Is that a problem? I don't think so. It is just SMOPs (Small Matter Of Programming)

> 
> > it doesn't matter either as we do not, and never will have an authorization from GC to get this data and use it in our own application.
> 
> If that's the assumption we can stop the whole discussion about "do we want to implement Labs API" right here?!

We will continue to have that discussion but not here on this channel. But to be clear: The discussion is not about the "Labs API", aka "public API" which is only a temporary solution, which we're allowed to use until we're ready to consume what is known as the "partner API". The key we're using for the "Labs API" will inevitably expire, sooner rather than later. So the real discussion is about partnering with GC by consuming their "partner API", aka "Live API", which happens to include some rather boring adventure lab methods which give us less than we already have.

**Comment by baiti** [comment]
Since this issue is subject to review, I am updating it:

I believe this issue is fixed with PR [#14318](https://github.com/cgeo/cgeo/pull/14318)

The typical scenario is that you solve a Lab Cache with the Lab Player App and once you did and refresh that cache in cgeo you will get the true found state and date, just like for any other cache. There is no need to configure setting the found sate manually except in cases where a user uses a different account to play lab caches in an attempt not to pollute his statistics with inflationary lab caches. In that case, using his premium account on cgeo he wants the capability to manyally synchronize the found sate with the account he uses to play the lab game.

**Comment by moving-bits** [comment]
I'm using current release 2025.04.02, and here no found dates are displayed for lab adventures, although I have updated them (to get the true found state).

**Comment by baiti** [comment]
> I'm using current release 2025.04.02, and here no found dates are displayed for lab adventures, although I have updated them (to get the true found state).

I have it a brief test. I have refreshed one that I now I completed and I can get the true found state, I see the badge on the icon and "Status Found" on the "Details" screen but indeed, no date. I need to investigate what's happening.

I believe it should show "Status Found (xx/yy/zz)", right?

**Comment by moving-bits** [comment]
> I have it a brief test. I have refreshed one that I now I completed and I can get the true found state, I see the badge on the icon and "Status Found" on the "Details" screen but indeed, no date.

Same here

> I believe it should show "Status Found (xx/yy/zz)", right?

Yes

**Comment by baiti** [comment]
Even though the issue ist titled "Support found date ...", the thread exlcusively discusses "Found state" and that we got a solution for since quite a while. 

Did we ever have a "true found date", e.g. the date a lab cache was completed using "the app", a property we can later retrieve via the Lab API? In other words: did this ever work?

**Comment by baiti** [comment]
I have just checked what we get from the API. There is no overall `CompletedDate` for  the lab cache as a whole, however, there is a `CompletedDate` per each what we call "Stage" and we know the number of stages so we could iterate them all and if all have a `CompletedDate`, we could take the latest such date and elevate it to the `CompletedDate` at the lab cache object level.

Would you think that this makes sense? I could give it a try then.

But first lets find out whether we had that functionality "ever"? If we had, we'd be dealing with a bug rather than a new feature.

**Comment by moving-bits** [comment]
I can't remember having seen it at all, but it would be nice to have that info. Maybe even on a per-stage level.

**Comment by ztNFny** [comment]
There never was a global found date in the lab API, so if the "use the highest stage date" code isn't there then it never was.

**Comment by baiti** [comment]
> There never was a global found date in the lab API, so if the "use the highest stage date" code isn't there then it never was.

That is what I suspected too. The "use highest stage date" code is not there yet, it was an idea I came up with yesterday. I will implement this as time permits.

When I do this, of course it is possible to render the date/time found on a per stage level too. Depending on how we do this, for instance prepending or appending it to some existing string, or define a dedicated "field" for this, it will require some UI work.

**Comment by baiti** [comment]
I have done some investigations and here is the result: In order to get the foundDate we need to make an API request to
`https://labs-api.geocaching.com/Api/GeocacheLogs` (aka a user's journal) which not only requires a `consumer_key` but also an `access_token` which must be acquired by an explicit login service or some Oauth2 flow which I have so far not been able to trace. I can say, however, that the explicit login is not possible with our `cgeo consumer_key`. So at this time I see no way to get to the completion dates as none of the API requests we're making, gives us any date/time, neither for the adventure itself nor for the stages solved.

@ztNFny : do you have some additional insights or ideas here?

**Comment by ztNFny** [comment]
Nope, not possible without retrieving either all the labs or the owners logs.
